### PR TITLE
Features/feature skipped question

### LINKED ISSUE #81: [BUG] Skip Question button functionality
**Describe the bug**
In the exam mode the `Skip Question` button is not working properly

**To Reproduce**
1. Click on Exam Mode
2. Click on Begin Exam
3. Click on Skip Question until no more questions left

**Expected behavior**
When the user clicks on the Skip Question button, the current question should be added to a queue of skipped questions. 
The questions from the queue should be showed to the user after all the questions from the initial queue has been answered.
The exam should not end until the user has answered all the questions, including the ones in the queue, or the time limit has expired.

### PR DESCRIPTION
**Pull request by Amnish04**
Fixes #81 

- [x] I have read and agree to the [Code of Conduct](https://github.com/eduardconstantin/azure-fundamentals/blob/main/CODE_OF_CONDUCT.md), and have followed the [Contributing guidelines](https://github.com/eduardconstantin/azure-fundamentals/blob/main/CONTRIBUTING.md).

## Pull Request Type

- [x] Bug fix (please link related issues if applicable)
- [x] New feature
- [ ] Question update
- [ ] Other

## Summary

@eduardconstantin  
Just finished with my implementation of requested functionality. Here is what I added:
- Added the implementation of a basic FIFO queue to be used in the actual feature.
- Any skipped questions are added to a `skippedQueue` saved in component state.
- Once the user goes over all the questions, the state is marked as `allQuestionsTouched`, and the user is brought to the first skipped question.
- Now every time user clicks next, he/she is taked to the next skipped question while the current question is removed from queue if answered.
- If the user decides to skip an already `skipped` question, they are brought back to it once they complete their current cycle of skipped questions.
- Fixed an **existing bug**, where `next button` was disabled in any scenario. Now, the next button only gets enabled once the user selects an option.

Hope you find these changes useful. Please let me know if something needs to be changed.

Thanks!

### CODE REVIEW

File: utils/Queue.ts
Comment (eduardconstantin) [inline utils/Queue.ts:1]: Could you replace the class with function?

**Review by eduardconstantin — CHANGES_REQUESTED** [review]

**Comment by eduardconstantin** [comment]
I also found a bug, if you select one option and then skip the question, when the question appears again the option remains selected.
image

You can fix this or create an issue for it.

File: utils/Queue.ts
Comment (Amnish04) [reply utils/Queue.ts:1]: Just finished with latest imlementation.

**Comment by Amnish04** [comment]
@eduardconstantin Just finished with all the changes requested.
1. Switched from a class to a functional style for `Queue` data structure.
2. Fixed the bug where the skipped questions preserve any selection. (I feel like that was a feature instead of a bug)
3. Resolved merge conflicts with latest changes in dev.

Hope it addresses the issues.

File: pages/exam/index.tsx [hunk 1]
```diff
@@ -67,16 +70,41 @@ const Exam: NextPage = () => {
...
   };
 
+  const handleSkipQuestion = (questionNo: number) => {
+    skippedQuestions.enqueue(questionNo);
+
+    // Unset any selected answer
+    setAnswers(prevState => {
+      if (prevState[questionNo]) delete prevState[questionNo];
+      return prevState;
+    });
```
Comment (masterHAWK99) [inline pages/exam/index.tsx:103]: You shouldn't modify the original object from state, instead, create a copy of this object, delete key, and finally update the state with a new object.

**Review by masterHAWK99 — CHANGES_REQUESTED** [review]

File: components/ExamQuizForm.tsx [hunk 2]
```diff
@@ -58,14 +62,21 @@ const ExamQuizForm: React.FC<Props> = ({
...
       </ul>
       <div className="flex justify-center flex-col sm:flex-row gap-4">
-        <Button type="submit" intent="primary" size="medium">
+        <Button 
+        type="button" 
+        intent="primary" 
+        size="medium"
+        onClick={() => {
+          reset(); handleSkipQuestion(currentQuestionIndex);
+        }}
```
Comment (masterHAWK99) [inline components/ExamQuizForm.tsx:71]: Please reformat your code manually, as you may have forgotten to run `npm run prepare`, causing it to not format while being committed.

**Review by masterHAWK99 — CHANGES_REQUESTED** [review]

File: pages/exam/index.tsx (hunk 1 above)
Comment (Amnish04) [reply pages/exam/index.tsx]: Fixed in the latest commit. Thanks for pointing out

File: components/ExamQuizForm.tsx (hunk 2 above)
Comment (Amnish04) [reply components/ExamQuizForm.tsx]: You're right, I skipped that part. Should be fixed in the latest commit

File: components/ExamQuizForm.tsx [hunk 3]
```diff
@@ -8,18 +8,22 @@ type Props = {
...
   questionSet,
   handleNextQuestion,
+  handleSkipQuestion,
   currentQuestionIndex,
   totalQuestions,
   setAnswer,
+  isQuestionAnswered,
 }) => {
   const { register, handleSubmit, reset } = useForm();
   const [lastIndex, setLastIndex] = useState<number>(0);
```
[the remark this answers is not in the document]
Comment (masterHAWK99) [reply components/ExamQuizForm.tsx:29]: I meant `const [lastIndex, setLastIndex] = useState<number>(0);`

[the remark this answers is not in the document]
Comment (Amnish04) [reply components/ExamQuizForm.tsx:29]: I see. To be honest, I am not sure why it was there in the first place. So I wouldn't prefer removing something I didn't add.

[Do not remove a fence until you know why it was put up in the first place.](https://fs.blog/chestertons-fence/)

**Comment by eduardconstantin** [comment]
> I believe that in the top left corner, there should be a display of the number of answered questions. Currently, when we skip some questions and the test ends, it shows the index number of the skipped question. For example, if I skipped the 3rd question and answered the remaining questions, when the test ended, the display in the top left corner showed "3/30" instead of "29/30". This may cause confusion.
> 
> What do you think @eduardconstantin ?

Either that or the skipped questions should not be counted. I tested with 4 questions, I skipped 2 and answered 2, then the counter got resetted, I answered the 2 skipped questions and got 2/4 question answered, which is wrong, I answered all 4.
image

File: components/ExamQuizForm.tsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (eduardconstantin) [reply components/ExamQuizForm.tsx:29]: @masterHAWK99 this is yours
image

**Comment by Amnish04** [comment]
> > I believe that in the top left corner, there should be a display of the number of answered questions. Currently, when we skip some questions and the test ends, it shows the index number of the skipped question. For example, if I skipped the 3rd question and answered the remaining questions, when the test ended, the display in the top left corner showed "3/30" instead of "29/30". This may cause confusion.
> > What do you think @eduardconstantin ?
> 
> Either that or the skipped questions should not be counted. I tested with 4 questions, I skipped 2 and answered 2, then the counter got resetted, I answered the 2 skipped questions and got 2/4 question answered, which is wrong, I answered all 4. image

@eduardconstantin 
Yes, currently it shows the index of the skipped question you're on.
In other words, you're on ith question out of n questions (i/n).

One thing we can do is update the questions counter when all skipped questions have been answered, like in normal case with no skipped questions.
That would look like `n/n`.

**Comment by Amnish04** [comment]
That is due to the way application has been set up. It is using an array of questions as the source `data.randomQuestions`, and updates the current index to switch to a question.
The counter above, based on the existing logic, does not represent the number of answered questions but what question we're currently at. And when you're done with the inital cycle of questions, you're sequentially taken back to the skipped questions. For example, you skipped question x, then y.

**Comment by Amnish04** [comment]
@eduardconstantin Also, to address the issue of knowing current question number, I have added that before a question's text.

image

**Comment by eduardconstantin** [comment]
> @eduardconstantin Also, to address the issue of knowing current question number, I have added that before a question's text.
> 
> image
> 
> Just pushed

This is not a bad idea

**Comment by Amnish04** [comment]
> > @eduardconstantin Also, to address the issue of knowing current question number, I have added that before a question's text.
> > image
> > Just pushed
> 
> This is not a bad idea

The changes are already in!

**Comment by eduardconstantin** [comment]
I tested and it look much better!
One thing though, when I select an option it gets counted as an answered question, shouldn't be counted when I press the next question button?
I can merge it like this and you can create an issue for this bug or you can fix it in this PR, it's up to you.

**Comment by Amnish04** [comment]
> I tested and it look much better! One thing though, when I select an option it gets counted as an answered question, shouldn't be counted when I press the next question button? I can merge it like this and you can create an issue for this bug or you can fix it in this PR, it's up to you.

That is because the current logic is setting the answers object, whenever a user clicks on an option instead of next button click handler. I think it belongs to a separate issue, will open it for you.

**Review by eduardconstantin — APPROVED** [review]

**Comment by Amnish04** [comment]
Great, just one last thing. Could you also add `hacktoberfest-accepted` label to this PR. Thanks!

**Comment by eduardconstantin** [comment]
I added the hacktoberfest label, it should count

**Pull request merged**